lib/rack/utils.rb

```diff
@@ -374,14 +374,22 @@ def get_byte_ranges(http_range, size)
     # that have already been processed by HMAC. This should not be used
     # on variable length plaintext strings because it could leak length info
     # via timing attacks.
-    def secure_compare(a, b)
-      return false unless a.bytesize == b.bytesize
+    if defined?(OpenSSL.fixed_length_secure_compare)
+      def secure_compare(a, b)
+        return false unless a.bytesize == b.bytesize
 
-      l = a.unpack("C*")
+        OpenSSL.fixed_length_secure_compare(a, b)
+      end
```

Comment:
What is the point to use `fixed_length_secure_compare` but check string lengths in prior? The comment above reads:
> NOTE:  the values compared should be of fixed length, such as strings
>    # that have already been processed by HMAC. This should not be used
>    # on variable length plaintext strings because it could leak length info
>    # via timing attacks.

When we fail fast on length difference this clearly opens the possibilities to timing attacks when this method is used inappropriately. And there is no way for a project to know that method was used inappropriately because it just returns `false`.

If you read description of [OpenSSL.fixed_length_secure_compare](https://www.rubydoc.info/stdlib/openssl/OpenSSL.fixed_length_secure_compare) you will see that it just raises if arguments are not same length. To show that its usage is unsafe.

If we believe that this method is widely misused and making it raise will cause too much harm, then just use `OpenSSL.secure_compare` that properly handles variable length string, just slower. Then implement a new rack method for fixed length secure compare that will make it clear by its name that only same length strings are compared. And that new method would behave as the  [OpenSSL.fixed_length_secure_compare](https://www.rubydoc.info/stdlib/openssl/OpenSSL.fixed_length_secure_compare) method that raises when lengths did not match.

Current implementation just overrides the mindful protection mechanism of [OpenSSL.fixed_length_secure_compare](https://www.rubydoc.info/stdlib/openssl/OpenSSL.fixed_length_secure_compare).

Reply:
In the current case (as in the case before this PR), there is timing attack that could allow an attacker to determine the number of bytes we are comparing against.  However, assuming that that number is not small (16 bytes or more), it seems unlikely that the code is actually vulnerable.  Making this raise instead of returning false for a mismatched size would break backwards compatibility.